src/src/dynpower.cpp

```diff
@@ -0,0 +1,204 @@
+    return;
+  dynpower_last_linkstats_millis = now;
+
+  // =============  LQ-based power boost up ==============
+  // Quick boost up of power when detected any emergency LQ drops.
+  // It should be useful for bando or sudden lost of LoS cases.
+  uint32_t lq_current = CRSF::LinkStatistics.uplink_Link_quality;
+  uint32_t lq_avg = dynpower_mavg_lq;
+  int32_t lq_diff = lq_avg - lq_current;
+  dynpower_mavg_lq.add(lq_current);
+  // if LQ drops quickly (DYNPOWER_LQ_BOOST_THRESH_DIFF) or critically low below DYNPOWER_LQ_BOOST_THRESH_MIN, immediately boost to the configured max power.
+  if (lq_diff >= DYNPOWER_LQ_BOOST_THRESH_DIFF || lq_current <= DYNPOWER_LQ_BOOST_THRESH_MIN)
+  {
+      DynamicPower_SetToConfigPower();
+      return;
+  }
+
+  PowerLevels_e startPowerLevel = POWERMGNT::currPower();
+  if (ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshUp == DYNPOWER_SNR_THRESH_NONE)
+  {
+    // =============  RSSI-based power increment ==============
+    // a simple threshold compared against N sample average of
+    // rssi vs the sensitivity limit +/- some thresholds
+    dynpower_mean_rssi.add(rssi);
+
+    if (dynpower_mean_rssi.getCount() >= DYNPOWER_RSSI_CNT)
+    {
+      int32_t expected_RXsensitivity = ExpressLRS_currAirRate_RFperfParams->RXsensitivity;
+      int8_t rssi_inc_threshold = expected_RXsensitivity + DYNPOWER_RSSI_THRESH_UP;
+      int8_t rssi_dec_threshold = expected_RXsensitivity + DYNPOWER_RSSI_THRESH_DN;
+      int8_t avg_rssi = dynpower_mean_rssi.mean(); // resets it too
+      if ((avg_rssi < rssi_inc_threshold) && (powerHeadroom > 0))
+      {
+        DBGLN("+power (rssi)");
+        POWERMGNT::incPower();
+      }
+      else if (avg_rssi > rssi_dec_threshold && lq_avg >= DYNPOWER_LQ_THRESH_DN)
+      {
+        DBGVLN("-power (rssi)");
+        POWERMGNT::decPower();
+      }
+    }
+  } // ^^ if RSSI-based
+  else
+  {
+    // =============  SNR-based power increment ==============
+    // Decrease the power if SNR above threshold and LQ is good
+    // Increase the power for each (X) SNR below the threshold
+    //int8_t snrScaled = CRSF::LinkStatistics.uplink_SNR;
+    if (snrScaled >= ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshDn && lq_avg >= DYNPOWER_LQ_THRESH_DN)
+    {
+      DBGVLN("-power (snr)");
+      POWERMGNT::decPower();
+    }
+
+    while ((snrScaled <= ExpressLRS_currAirRate_RFperfParams->DynpowerSnrThreshUp) && (powerHeadroom > 0))
+    {
+      DBGLN("+power (snr)");
+      POWERMGNT::incPower();
+      // Every power doubling will theoretically increase the SNR by 3dB, but closer to 2dB in testing
+      snrScaled += SNR_SCALE(2);
+      --powerHeadroom;
+    }
+  } // ^^ if SNR-based
+
+  // If instant LQ is low, but the SNR/RSSI did nothing, inc power by one step
+  if ((powerHeadroom > 0) && (startPowerLevel == POWERMGNT::currPower()) && (lq_current <= DYNPOWER_LQ_THRESH_UP))
+  {
+    DBGLN("+power (lq)");
+    POWERMGNT::incPower();
+    //--powerHeadroom;
```

Comment:
Shouldn't this be uncommented?

Reply:
No it is commented and left in so if any code is added after it, it will have to be returned. It is commented out just in case the complier decides it actually wants to leave that code in and costs us 4 bytes. Every byte is sacred!